### DISCUSSION TITLE
to numerate shapes

### DISCUSSION THREAD

**Discussion by 007CPS** [spakin/SimpInkScr#129] [Q&A]
Hello,

I would like to numerate each shape of a selection. 
I found a solution by exploring the parameters:

```
print(selected_shapes())

all_selected_shape = selected_shapes()
print(all_selected_shape[0])
print(all_selected_shape[0].__dict__)

print(all_selected_shape[0]._inkscape_obj)

print(all_selected_shape[0]._transform)

print(type(all_selected_shape[0]._transform.matrix))
print(all_selected_shape[0]._transform.matrix)

print(all_selected_shape[0]._transform.matrix[0])

print(type(all_selected_shape[0]._transform.matrix[0][0]))
print(all_selected_shape[0]._transform.matrix[0][0])

print("positionX:", all_selected_shape[0]._transform.matrix[0][2])
print("positionY:", all_selected_shape[0]._transform.matrix[1][2])
```
output:
![4](https://github.com/spakin/SimpInkScr/assets/109821460/e23f1189-e36b-4917-9640-d98f4c37ee9b)

My actual solution is to obtain the coordinates from ._transfrom.
My code is as follows:

```
   all_selected_shape = selected_shapes()
   n_id = 1
   
   for obj in all_selected_shape:
       obj_x = obj._transform.matrix[0][2]
       obj_y = obj._transform.matrix[1][2]
       
       t = text(f'{n_id}', (obj_x, obj_y), font_size='10pt')
       print(n_id,obj_x,obj_y)
       n_id=n_id+1
```

In the example of a selected triangle, we have the following coordinates:
![3](https://github.com/spakin/SimpInkScr/assets/109821460/d61104d5-292e-43b4-bf67-0f798dea43cd)

the result is shown in the following image:
![2](https://github.com/spakin/SimpInkScr/assets/109821460/8b23c5b1-35dc-4c34-ac92-7ef8d0b92881)

I guess that the numering is done on the first node.

How can i get correctly the coordinates (in another way than .transform) and on the center of the shape ?

Thanks!

**Comment by spakin** [comment]
It's possible to iterate over the control points of the underlying `inkex` object.  The center can be computed either manually (by averaging all control points) or by asking the shape's bounding box for its center.

Here's some code to get you started:
```Python
for obj in selected_shapes():
    if not isinstance(obj, SimplePathObject):
        continue
    iobj = obj.get_inkex_object()
    for i, ep in enumerate(iobj.get_path().end_points):
        text(str(i + 1), ep, font_size='10pt')
    text('C', iobj.bounding_box().center, font_size='10pt')
```

**Reply by 007CPS** [reply]
Thanks for your solution. I tried and it works for wide triangle.
![Capture](https://github.com/spakin/SimpInkScr/assets/109821460/aa3ed426-096e-470c-807f-ac0acccdf4cf)

but i don't understand why there are several text layers:
![Capture2](https://github.com/spakin/SimpInkScr/assets/109821460/bd4aba62-a032-432b-8ade-7f9efaa12f86)

Plus, if we take into account thin polygonal shape, it doesn't really work any more
![Capture3](https://github.com/spakin/SimpInkScr/assets/109821460/3d6fb4d0-78f4-4387-bbca-cb8fc63d7be6)

How can we aske for the shape's bounding box?

**Reply by spakin** [reply]
> but i don't understand why there are several text layers:

There aren't; Inkscape is showing you the individual objects within the layer.  (That's why the panel is called "Layers *and Objects*".)  You can turn that off—I usually do—by clicking the gear icon at the top of the panel and checking _Only show layers_.

> Plus, if we take into account thin polygonal shape, it doesn't really work any more

What's not working?  In your screenshot, C is indeed located at the center of the bounding box containing the thin polygonal shape.  For any other definition of "center" you'll need to calculate the point yourself.  For example, the following code draws "Ca" at the average endpoint coordinate and, for comparison, "Cb" at the center of the bounding box:

```Python
for obj in selected_shapes():
    if not isinstance(obj, SimplePathObject):
        continue
    iobj = obj.get_inkex_object()
    average = inkex.Vector2d(0, 0)
    endpoints = list(iobj.get_path().end_points)
    for i, ep in enumerate(endpoints):
        text(str(i + 1), ep, font_size='10pt')
        average += ep
    average /= len(endpoints)
    text('Ca', average, font_size='10pt')
    text('Cb', iobj.bounding_box().center, font_size='10pt')
```

Note that "center" is not a uniquely defined geometric property.  There are _tens of thousands_ of different centers of triangles according to the [Encyclopedia of Triangle Centers](https://faculty.evansville.edu/ck6/encyclopedia/ETC.html).  You'll have to think hard about what center you're actually looking for then find/write the Python code to compute it.
